lib/async_hooks.js

```diff
@@ -190,6 +191,123 @@ class AsyncResource {
 }
 
 
+// AsyncLocal //
+
+const kStack = Symbol('stack');
+const kIsFirst = Symbol('is-first');
+const kMap = Symbol('map');
+const kOnChangedCb = Symbol('on-changed-cb');
+const kHooks = Symbol('hooks');
+const kSet = Symbol('set');
+
+class AsyncLocal {
+  constructor(options = {}) {
+    if (typeof options !== 'object' || options === null)
+      throw new ERR_INVALID_ARG_TYPE('options', 'Object', options);
+
+    const { onChangedCb = null } = options;
+    if (onChangedCb !== null && typeof onChangedCb !== 'function')
+      throw new ERR_INVALID_ARG_TYPE('options.onChangedCb',
+                                     'function',
+                                     onChangedCb);
+
+    this[kOnChangedCb] = onChangedCb;
+    this[kMap] = new Map();
+
+    const fns = {
+      init: (asyncId, type, triggerAsyncId, resource) => {
+        // Propagate value from current id to new (execution graph)
+        const value = this[kMap].get(executionAsyncId());
+        if (value)
+          this[kMap].set(asyncId, value);
+      },
+
+      destroy: (asyncId) => this[kSet](asyncId, null),
+    };
+
+    if (this[kOnChangedCb]) {
+      // Change notification requires to keep a stack of async local values
+      this[kStack] = [];
+      // Indicates that first value was stored (before callback "missing")
+      this[kIsFirst] = true;
+
+      // Use before/after hooks to signal changes because of execution
+      fns.before = (asyncId) => {
+        const stack = this[kStack];
+        const cVal = this[kMap].get(asyncId);
+        const pVal = stack[stack.length - 1];
+        stack.push(pVal);
+        if (cVal !== pVal)
+          this[kOnChangedCb](pVal, cVal, true);
```

Comment:
This invocation is very risky. User callbask should be called outside of async hooks.

Reply:
Well, then it will not work as moving it outside of async hooks would require something like nextTick which is a new async operation. As a result the signaled values and the real values don't match anymore.

Reply:
If this cannot be made safe, I don't think supporting this "onchanged" behavior is compatible for the goal of "ease of use" that this API is supposed to have.

Reply:
I will improve docs in this, add a check to catch recursion cases (e.g. onChangedCb sets a new value) and handle exceptions thrown by onChangedCb. Hopefully this renders this callback as save as all the other callbacks which can be registered in node core.